### PR TITLE
[Serialization] Encode depth for cross-refs to generic parameters

### PR DESCRIPTION
**Pull request by jrose-apple**
Otherwise, we can't represent a cross-reference to generic parameters in a parent type *when used in an extension*.

[SR-9084](https://bugs.swift.org/browse/SR-9084) / rdar://problem/45566043

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
Oops, this…doesn't actually solve the problem yet. Please hold.

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
@swift-ci Please test compiler performance

File: lib/Serialization/Deserialization.cpp [hunk 1]
```diff
@@ -1660,6 +1660,18 @@ ModuleFile::resolveCrossReference(ModuleDecl *baseModule, uint32_t pathLen) {
...
+
+      unsigned currentDepth = paramList->getDepth();
+      if (currentDepth < depth) {
+        return llvm::make_error<XRefError>(
+            "a containing type has been made non-generic",
+            pathTrace, getXRefDeclNameForError());
+      }
+      while (currentDepth > depth) {
+        paramList = paramList->getOuterParameters();
+        --currentDepth;
```
Comment (slavapestov) [inline lib/Serialization/Deserialization.cpp]: Is it worth asserting that the GenericTypeParamDecl you find below has the right depth and index?

**Review by slavapestov — APPROVED** [review]

File: lib/Serialization/Deserialization.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/Serialization/Deserialization.cpp]: That seems like a good idea for the AST verifier in general. It's set up by construction that the depth matches the GenericParamList's notion of depth, but the GenericParamList and the GenericTypeParamDecl could be out of sync. I'll do that in a follow-up PR.

**Comment by jrose-apple** [comment]
These compiler perf numbers are all wrong again. @graydon, worth investigating, or should I just move on?

**Comment by graydon** [comment]
@jrose-apple I've pushed a fix that ought to give better detection of failed modules. If you don't mind, I'd like to run the same cperf test again?

**Comment by graydon** [comment]
@swift-ci Please test compiler performance

**Comment by jrose-apple** [comment]
Hm, not quite there yet…

**Comment by graydon** [comment]
@jrose-apple it's trying to figure out which modules failed in _either_ instance (old or new) so it can exclude them from comparison. I .. do not understand what happened here. Sigh. Will try again.

**Comment by graydon** [comment]
@jrose-apple one more time!

**Comment by graydon** [comment]
@swift-ci Please test compiler performance

**Comment by shahmishal** [comment]
@jrose-apple @graydon Going to have to kill compiler performance job, currently updating Xcode.

**Comment by jrose-apple** [comment]
@swift-ci Please test compiler performance

**Comment by jrose-apple** [comment]
@swift-ci Please test compiler performance

**Comment by jrose-apple** [comment]
@swift-ci Please test compiler performance

**Comment by jrose-apple** [comment]
There we go! Merging.

**Comment by graydon** [comment]
@jrose-apple thanks for your perseverance, I know it's tedious and in a better world we'd have set all this up to be much more directly testable :(

**Pull request merged**